src/main/java/frc/robot/subsystems/Elevator.java

```diff
@@ -0,0 +1,183 @@
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.DutyCycleEncoder;
+import edu.wpi.first.wpilibj.shuffleboard.ShuffleboardTab;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.constants.Constants;
+import frc.robot.constants.ElevatorConstants;
+import frc.robot.util.MotorFactory;
+import lib.ctre_shims.TalonEncoder;
+
+public class Elevator extends SubsystemBase {
+
+  private final WPI_TalonFX m_motor;
+  private final PIDController m_elevatorPID;
+  private final ElevatorFeedforward m_elevatorFF;
+  private final TalonEncoder m_talonEncoder; 
+  private final DigitalInput m_topLimitSwitch; 
+  private final DigitalInput m_bottomLimitSwitch; 
+  private final ShuffleboardTab m_elevatorTab; 
+  private double m_maxHeight; 
+  private boolean m_enabled; 
+  private boolean m_isCalibrated = false; 
+
+  public Elevator(ShuffleboardTab elevatorTab) {
+    m_elevatorTab = elevatorTab; 
+    m_motor = MotorFactory.createTalonFX(ElevatorConstants.kMotorPort, Constants.kCanivoreCAN);
+    m_motor.setInverted(true);
+    m_motor.setNeutralMode(NeutralMode.Brake);
+    m_talonEncoder = new TalonEncoder(m_motor); 
+    m_talonEncoder.setDistancePerPulse(ElevatorConstants.kDistPerPulse);
+    addChild("motor", m_motor);
+   
+    m_elevatorPID = new PIDController(ElevatorConstants.kP, ElevatorConstants.kI, ElevatorConstants.kD);
+    //m_elevatorTab.add("pid",m_elevatorPID);
+    //m_elevatorTab.addNumber("",()-> getHeight());
+    m_elevatorPID.setTolerance(0.03);
+    m_elevatorFF = new ElevatorFeedforward(ElevatorConstants.kS, ElevatorConstants.kG, ElevatorConstants.kV, ElevatorConstants.kA);
+
+    m_topLimitSwitch = new DigitalInput(ElevatorConstants.kTopLimitSwitchPort); 
+    m_bottomLimitSwitch = new DigitalInput(ElevatorConstants.kBottomLimitSwitchPort); 
+    addChild("Upper Limit",m_topLimitSwitch);
+    addChild("Bottom Limit",m_bottomLimitSwitch);
+
+
+    //TODO: log, addDouble doesn't work. 
+    //LogManager.addDouble("Elevator/error", () -> {return m_pid.getSetpoint() - getElevatorHeight();});
+    
+    //m_motor.setSafetyEnabled(true);
+    setUpElevatorTab();
+    setSetpoint(0.5); 
+  }
+
+  @Override
+  public void periodic() {
+    /**
+     * If we hit the bottom limit switch, then we must be at the zero position.
+     * Thus set the setpoint to 0, then atSetpoint() will be triggered, causing the 
+     * ResetEncoderAtBottom() command to end
+     */
+    if(m_enabled && m_isCalibrated) {
+      double pid = m_elevatorPID.calculate(getHeight());
+      double ff = m_elevatorFF.calculate(ElevatorConstants.kVelocity, ElevatorConstants.kAccel); 
+      ff = 0; 
+      double pidClamped = MathUtil.clamp(pid, -ElevatorConstants.kPowerLimit, ElevatorConstants.kPowerLimit);
+      double finalMotorPower = pidClamped + ff; 
```

Comment:
you should clamp after FF and dosen't evenlook like you use feed forward. If you are going to use FF note that you are not useing the calculate method correctly

Reply:
I would recommend just using the gravity term of the FF. The other terms would require trapezoid profile or something else

Reply:
yeah I am not sure about what is happening with FF as of now. Some testing is required to see if it is even needed. But point noted that I am calculating it wrong.

Reply:
Thank you for your feedback as well Anthony